patches/tModLoader/Terraria/ModLoader/UI/ModBrowser/UIModBrowser.cs

```diff
@@ -184,103 +236,139 @@ public void BackClick(UIMouseEvent evt, UIElement listeningElement)
-			ModList.Add(NoModsFoundText);
-		else if (hasNoModsFoundNotif)
-			ModList.RemoveChild(NoModsFoundText);
-		_rootElement.RemoveChild(_updateAllButton);
-		if (SpecialModPackFilter == null && _items.Count(x => x.ModDownload.HasUpdate && !x.ModDownload.UpdateIsDowngrade) > 0) _rootElement.Append(_updateAllButton);
+		_browserStatus.SetCurrentState(state);
+		_reloadButton.SetText(Language.GetText("tModLoader.MBGettingData"));
+	}
+	private void ModListFinished(AsyncProviderState state)
+	{
+		_browserStatus.SetCurrentState(state);
+		_reloadButton.SetText(Language.GetText("tModLoader.MBReloadBrowser"));
 	}
 
 	public override void OnActivate()
 	{
+		base.OnActivate();
 		Main.clrInput();
-		if (!Loading && _items.Count <= 0) {
-			if (WorkshopHelper.QueryHelper.Items.Count == 0)
-				PopulateModBrowser(uiOnly: false);
-			else
-				PopulateModBrowser(uiOnly: true);
+		if (_firstLoad) {
+			SocialBackend.Initialize(); // Note this is currently synchronous
+			PopulateModBrowser();
 		}
+
+		// Check for mods to update
+		// @NOTE: Now it's done only once on load
+		CheckIfAnyModUpdateIsAvailable();
+
+		DebounceTimer = null;
 	}
 
-	internal void PopulateModBrowser(bool uiOnly)
+	private void CbLocalModsChanged(HashSet<string> modSlugs)
 	{
-		// Initialize
-		Loading = true;
-		SpecialModPackFilter = null;
-		SpecialModPackFilterTitle = null;
-		_reloadButton.SetText(Language.GetText("tModLoader.MBGettingData"));
-		_backgroundElement.Append(_loaderElement);
-		SetHeading(Language.GetText("tModLoader.MenuModBrowser"));
+		// Can be called outside main thread
+		lock (modSlugsToUpdateInstallInfo) {
+			modSlugsToUpdateInstallInfo.UnionWith(modSlugs);
+		}
+		// Updates the 'Update All' button using cached data from GetInstalledModDownloadedItems
+		CheckIfAnyModUpdateIsAvailable();
+	}
 
-		// Remove old data
-		_items.Clear();
-		
-		// Asynchronous load the Mod Browser
-		Task.Run(() => {
-			InnerPopulateModBrowser(uiOnly: uiOnly);
-			Loading = false;
-			_reloadButton.SetText(Language.GetText("tModLoader.MBReloadBrowser"));
-		});
+	public override void OnDeactivate()
+	{
+		DebounceTimer = null;
+		base.OnDeactivate();
 	}
 
-	internal bool InnerPopulateModBrowser(bool uiOnly)
+	public override void Update(GameTime gameTime)
 	{
-		if (!uiOnly && !WorkshopHelper.QueryHelper.QueryWorkshop())
-			return false;
+		base.Update(gameTime);
+
+		if (_updateAvailable) {
+			_updateAvailable = false;
+			Interface.updateMessage.SetMessage(_updateText);
+			Interface.updateMessage.SetGotoMenu(Interface.modBrowserID);
+			Interface.updateMessage.SetURL(_updateUrl);
+			Interface.updateMessage.SetAutoUpdateURL(_autoUpdateUrl);
+			Main.menuMode = Interface.updateMessageID;
+		}
+
+		lock (modSlugsToUpdateInstallInfo) {
+			foreach (var item in ModList.ReceivedItems.Where(
+				d => modSlugsToUpdateInstallInfo.Contains(d.ModDownload.ModName)
+			)) {
+				item.ModDownload.UpdateInstallState();
+				item.UpdateInstallInfo();
```

Comment:
Seems to be 50/50 on detecting if an item was installed in this.
Bigger mods seem to have more issues from observation